server/neptune/gateway/event/project_config_builder.go

```diff
@@ -0,0 +1,99 @@
+	"github.com/runatlantis/atlantis/server/logging"
+)
+
+// repoFetcher manages a cloned repo's workspace on disk for running commands.
+type repoFetcher interface {
+	Fetch(ctx context.Context, baseRepo models.Repo, sha string) (string, func(ctx context.Context, filePath string), error)
+}
+
+// hooksRunner runs preworkflow hooks for a given repository/commit
+type hooksRunner interface {
+	Run(repo models.Repo, repoDir string) error
+}
+
+// fileFetcher handles being able to identify and fetch the changed files per individual commit
+type fileFetcher interface {
+	GetModifiedFilesFromCommit(ctx context.Context, repo models.Repo, sha string, installationToken int64) ([]string, error)
+}
+
+// rootFinder determines which roots were modified in a given event.
+type rootFinder interface {
+	// DetermineRoots returns the list of roots that were modified
+	// based on modifiedFiles and the repo's config.
+	DetermineRoots(modifiedFiles []string, config valid.RepoCfg) ([]valid.Project, error)
+}
+
+// parserValidator config builds repo specific configurations
+type parserValidator interface {
+	HasRepoCfg(absRepoDir string) (bool, error)
+	ParseRepoCfg(absRepoDir string, globalCfg valid.GlobalCfg, repoID string) (valid.RepoCfg, error)
+}
+
+type RootConfigBuilder struct {
+	RepoFetcher     repoFetcher
+	HooksRunner     hooksRunner
+	ParserValidator parserValidator
+	RootFinder      rootFinder
+	FileFetcher     fileFetcher
+	GlobalCfg       valid.GlobalCfg
+	Logger          logging.Logger
+}
+
+func (b *RootConfigBuilder) Build(ctx context.Context, event Push) ([]*valid.MergedProjectCfg, error) {
+	// Generate a new filepath location and clone repo into it
+	repoDir, deleteFn, err := b.RepoFetcher.Fetch(ctx, event.Repo, event.Sha)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("creating temporary clone at path: %s", repoDir))
+	}
+	defer deleteFn(ctx, repoDir)
+
+	// Run pre-workflow hooks
+	err = b.HooksRunner.Run(event.Repo, repoDir)
+	if err != nil {
+		return nil, errors.Wrap(err, "running pre-workflow hooks")
+	}
+
+	// Fetch files modified in commit
+	modifiedFiles, err := b.FileFetcher.GetModifiedFilesFromCommit(ctx, event.Repo, event.Sha, event.InstallationToken)
+	if err != nil {
+		return nil, errors.Wrapf(err, "finding modified files: %s", modifiedFiles)
+	}
+
+	// Validate rep cfg file exists
+	hasRepoCfg, err := b.ParserValidator.HasRepoCfg(repoDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "looking for %s file in %q", config.AtlantisYAMLFilename, repoDir)
+	}
+	if !hasRepoCfg {
+		return nil, errors.New("repo cfg file does not exist")
+	}
+
+	// Parse repo configs into specific root configs (i.e. roots)
+	// TODO: rename project to roots
+	var mergedRootCfgs []*valid.MergedProjectCfg
+	repoCfg, err := b.ParserValidator.ParseRepoCfg(repoDir, b.GlobalCfg, event.Repo.ID())
+	if err != nil {
+		return nil, errors.Wrapf(err, "parsing %s", config.AtlantisYAMLFilename)
+	}
```

Comment:
Would be nice to coalesce this into just one function `ParseRepoCfg` which fails if there is no repo config and we just return that error instead of having another method just to check existence.

Reply:
we can, but logic is currently shared with pre-neptune way of doing this. To minimize changes to old workflow, I'd need to duplicate this logic in the gateway pkg like other dependencies. Will do so in a followup PR.

Reply:
Created a new ParserValidator implementation for gateway. Increased PR size is mainly from duplicating the relevant yaml validation tests from the older version. There are some things I removed/nits cleaned up/etc. but overall functionality should be the same. We just check if the config file exists within  `ParseRepoCfg()`.